drivers: watchdog: wdt_nrfx.c: Unistall timeouts in wdt_disable()

nordic-segl opened the pull request:
According to API description, wdt_disable() shall
automatically uninstall all timeouts.
https://docs.zephyrproject.org/latest/hardware/peripherals/watchdog.html

Implement missing wdt_disable() functionality:
- disable WDT RR channels;
- remove callback from timeout channels;
- set data->m_allocated_channels to zero.

nordic-segl committed:
drivers: watchdog: wdt_nrfx.c: Unistall timeouts in wdt_disable()

According to API description, wdt_disable() shall
automatically uninstall all timeouts.
https://docs.zephyrproject.org/latest/hardware/peripherals/watchdog.html

Implement missing wdt_disable() functionality:
- disable WDT RR channels;
- remove callback from timeout channels;
- set data->m_allocated_channels to zero.

Signed-off-by: Sebastian Głąb <sebastian.glab@nordicsemi.no>

nordic-segl commented:
DNM before hal_nordic supports WDT channel deallocation.

kl-cruz commented:
Is it now ready to work on?

nordic-segl commented:
> Is it now ready to work on?

No,
void nrfx_wdt_channels_free(nrfx_wdt_t const * p_instance);
is missing in hal_nordic. Shall I cherry pick that change from NRFX repo or is there a plan for rebase/new release?

ArekBalysNordic commented:
> > Is it now ready to work on?
> 
> No, void nrfx_wdt_channels_free(nrfx_wdt_t const * p_instance); is missing in hal_nordic. Shall I cherry pick that change from NRFX repo or is there a plan for rebase/new release?

If it is possible it would be great because I would like to have a fix for that as soon as possible. :)

nordic-segl commented:
> > > Is it now ready to work on?
> > 
> > 
> > No, void nrfx_wdt_channels_free(nrfx_wdt_t const * p_instance); is missing in hal_nordic. Shall I cherry pick that change from NRFX repo or is there a plan for rebase/new release?
> 
> If it is possible it would be great because I would like to have a fix for that as soon as possible. :)

It's already committed to the hal_nordic repo
https://github.com/zephyrproject-rtos/hal_nordic/commit/8eae3cc15977e4a232e3e5ca1088c048d4ceb263#diff-4e564f3f3dea92d458fc3a9d7c3c2bf6c3b0a6fd2ae9b6906476f6912b03fd10

Here is PR that brings in updated hal_nordic:
https://github.com/zephyrproject-rtos/zephyr/pull/72586

nordic-segl committed:
tests: drivers: watchdog: wdt_error_cases: Add test scenario

Add test scenario that verifies correct error code
returned by wdt_setup() when called just after wdt_disable().

Scenario:
- configure WDT timeout;
- start WDT by calling wdt_setup();
- remove all WDT timeouts by calling wdt_disable();
- confirm wdt_setup() returns error code due to lack
  of configured WDT timeouts.

Signed-off-by: Sebastian Głąb <sebastian.glab@nordicsemi.no>

nordic-segl commented on drivers/watchdog/wdt_nrfx.c:
```diff
@@ -69,6 +71,13 @@ static int wdt_nrf_disable(const struct device *dev)
+		data->m_callbacks[channel_id] = NULL;
+	}
+	data->m_allocated_channels = 0;
+
```
I add one more test to check if this left over can lead to WDT start without any timeouts configured.

nordic-segl replied:
I run "extended" wdt_error_cases manually on nrf54l15pdk and nrf54h20. All tests are passing (so fix in this PR is working as expected).
